pkg/storage/odf_test.go

```diff
@@ -0,0 +1,401 @@
+			client:        true,
+		},
+		{
+			name:          defaultStorageSystemName,
+			namespace:     "",
+			expectedError: "storageSystem 'nsname' cannot be empty",
+			client:        true,
+		},
+		{
+			name:          defaultStorageSystemName,
+			namespace:     defaultStorageSystemNamespace,
+			expectedError: "",
+			client:        false,
+		},
+	}
+
+	for _, testCase := range testCases {
+		var testSettings *clients.Settings
+
+		if testCase.client {
+			testSettings = clients.GetTestClients(clients.TestClientParams{})
+		}
+
+		testStorageSystemBuilder := NewStorageSystemBuilder(testSettings, testCase.name, testCase.namespace)
+
+		if testCase.expectedError == "" {
+			if testCase.client {
+				assert.Equal(t, testCase.name, testStorageSystemBuilder.Definition.Name)
+				assert.Equal(t, testCase.namespace, testStorageSystemBuilder.Definition.Namespace)
+			} else {
+				assert.Nil(t, testStorageSystemBuilder)
+			}
+		} else {
+			assert.Equal(t, testCase.expectedError, testStorageSystemBuilder.errorMsg)
+			assert.NotNil(t, testStorageSystemBuilder.Definition)
+		}
+	}
+}
+
+func TestStorageSystemExist(t *testing.T) {
+	testCases := []struct {
+		testStorageSystem *StorageSystemBuilder
+		expectedStatus    bool
+	}{
+		{
+			testStorageSystem: buildValidStorageSystemBuilder(buildStorageSystemClientWithDummyObject()),
+			expectedStatus:    true,
+		},
+		{
+			testStorageSystem: buildInValidStorageSystemBuilder(buildStorageSystemClientWithDummyObject()),
+			expectedStatus:    false,
+		},
+		{
+			testStorageSystem: buildValidStorageSystemBuilder(clients.GetTestClients(clients.TestClientParams{})),
+			expectedStatus:    false,
+		},
+	}
+
+	for _, testCase := range testCases {
+		exist := testCase.testStorageSystem.Exists()
+		assert.Equal(t, testCase.expectedStatus, exist)
+	}
+}
+
+func TestStorageSystemGet(t *testing.T) {
+	testCases := []struct {
+		testStorageSystem *StorageSystemBuilder
+		expectedError     error
+	}{
+		{
+			testStorageSystem: buildValidStorageSystemBuilder(buildStorageSystemClientWithDummyObject()),
+			expectedError:     nil,
+		},
+		{
+			testStorageSystem: buildInValidStorageSystemBuilder(buildStorageSystemClientWithDummyObject()),
+			expectedError:     fmt.Errorf("storageSystem 'name' cannot be empty"),
+		},
+		{
+			testStorageSystem: buildValidStorageSystemBuilder(clients.GetTestClients(clients.TestClientParams{})),
+			expectedError: fmt.Errorf("storagesystems.odf.openshift.io \"ocs-storagecluster-storagesystem\" " +
+				"not found"),
+		},
+	}
+
+	for _, testCase := range testCases {
+		storageSystemObj, err := testCase.testStorageSystem.Get()
+
+		if testCase.expectedError == nil {
+			assert.Equal(t, storageSystemObj.Name, testCase.testStorageSystem.Definition.Name)
+			assert.Equal(t, storageSystemObj.Namespace, testCase.testStorageSystem.Definition.Namespace)
+			assert.Nil(t, err)
+		} else {
+			assert.Equal(t, testCase.expectedError.Error(), err.Error())
```

Comment:
Can you move this above line 209 and always check it?

Reply:
@trewest please, let's leave it like this. It is imposible to use it generic. I need error message, I cannot compare error with error they are not the same. I need to verify error msg is as expected. But,when error is not expected == nil there is no object to get msg from and there is panic failure:
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x212c940]

Reply:
Panics should only occur if you call the Error() method on an error that is nil. It isn't necessary for determining error messages, you just have to compare the error objects directly `assert.Equal(t,  testCase.expectedError, err)`. Please let me know if that example doesn't work in your scenario

Reply:
in this specific case it is work, but I changed an expectedError type